fetchurl: reuse lib.normalizeHashes

balsoft opened the pull request:
- **fetchurl: re-use lib.normalizeHashes**

  Reuses the normalizeHashes function instead of reimplementing it.

  I've tested this manually and the behaviour seems to be the same.
- **normalizeHash: trace when a fake hash is assumed**
  Needed to keep the neat trace when only empty hashes are passed.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

balsoft committed:
normalizeHash: warn when a fake hash is assumed

This replicates the behaviour of `fetchurl` for other fetchers too,
except it `warn`s instead of `trace`ing, which seems more appropriate.

balsoft committed:
fetchurl: re-use lib.normalizeHashes

Re-uses normalizeHashes instead of reimplementing it in a verbose way.

balsoft commented on lib/fetchers.nix:
```diff
@@ -132,7 +133,10 @@ rec {
+            let
+              fakeHash = fakeH.${h.name} or (throw "no “fake hash” defined for ${h.name}");
+            in
+            trace "found empty ${h.name}, assuming '${fakeHash}'" fakeHash
```
It used to be `lib.trace` in `fetchurl`. I also agree that `lib.warn` would be a better fit, I'll use it.

balsoft commented on pkgs/build-support/fetchurl/default.nix:
```diff
@@ -144,54 +144,23 @@ let
-        outputHash = "";
-      }
-    else
-      throw "fetchurl requires a hash for fixed-output derivation: ${lib.generators.toPretty { } urls_}";
```
Hm, good catch. I will add it back in.

balsoft replied:
Sorry, I've realized I should refactor the code a bit to improve the error messaging for other fetchers too. Can you check it out again?

alyssais reviewed (approved):
Looks good to me, but I'd like somebody else to review as well, in case I've missed a behaviour change.

balsoft committed:
lib.normalizeHash: improve error reporting

Adds an optional `fetcher` argument that gets tacked on to error messages
and warnings for better error reporting and easier debugging.

alyssais commented on lib/fetchers.nix:
```diff
@@ -112,16 +118,19 @@ rec {
       args
     else
       let
+        hashTypes_ = unique ([ "hash" ] ++ hashTypes);
```
[nit] I'd find `hashTypes'` for idiomatic — it's basically what the `'` character in identifiers is there for.

infinisil reviewed (changes requested):
- Calling with just `outputHash` has a worse error message:
  ```nix
  fetchurl {
    url = "<url>";
    outputHash = "0000000000000000000000000000000000000000000000000000";
  }
  ```

  Before:
  ```
  error: fetchurl was passed outputHash without outputHashAlgo: [
         "<url>"
       ]
  ```

  After:
  ```
  error: attribute 'outputHashAlgo' missing
  ```
- Only passing `outputHashAlgo` now doesn't work anymore and gives a confusing error message:
  ```nix
  fetchurl { url = "<url>"; outputHashAlgo = "sha256"; }
  ```

  Before:
  ```
  warning: found empty hash, assuming 'sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA='
  ```

  After:
  ```
  error: hash 'sha256' does not include a type, nor is the type otherwise known from context
  ```

I might try implementing this better myself in a bit.

infinisil commented:
I also just noticed that the [performance numbers](https://github.com/NixOS/nixpkgs/actions/runs/18105228931?pr=444328) of this change are quite bad, such as:

- gc.totalBytes: +2.3165%
- nrThunks: +3.4213%
- nrFunctionCalls: +6.4420%
